core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -620,6 +620,14 @@ private[spark] class TaskSetManager(
     // Note: "result.value()" only deserializes the value when it's called at the first time, so
     // here "result.value()" just returns the value and won't block other threads.
     sched.dagScheduler.taskEnded(tasks(index), Success, result.value(), result.accumUpdates, info)
+    // Kill other task attempts if any as the one attempt succeeded
+    for (attemptInfo <- taskAttempts(index) if attemptInfo.attemptNumber != info.attemptNumber
```

Comment:
I can think that during the map phase(which don't write to Hadoop) there is a chance of succeeding two attempts as you explained. But in final phase(which write to Hadoop) tasks, during commitTask() if two attempts try to rename taskAttemptPath to committedTaskPath then only one attempt would succeed and other will fail with the rename failure.